src/storage/cache/rpc/packet.rs

```diff
@@ -0,0 +1,283 @@
+
+        Ok(ReqHeader { seq, op, len })
+    }
+}
+
+/// The message module contains the data structures shared between the client and server.
+/// Assume the cluster only contains one version server, so we won't consider the compatibility
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub struct RespHeader {
+    /// The sequence number of the response.
+    pub seq: u64,
+    /// The operation type of the response.
+    /// 0: keepalive
+    /// other is defined by the user
+    pub op: u8,
+    /// The length of the response.
+    pub len: u64,
+}
+
+impl Encode for RespHeader {
+    fn encode(&self) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(RESP_HEADER_SIZE as usize);
+        buf.extend_from_slice(&self.seq.to_be_bytes());
+        buf.push(self.op);
+        buf.extend_from_slice(&self.len.to_be_bytes());
+        buf
+    }
+}
+
+impl Decode for RespHeader {
+    fn decode(buf: &[u8]) -> Result<Self, RpcError<String>> {
+        if buf.len() < RESP_HEADER_SIZE as usize {
+            return Err(RpcError::InvalidResponse("Invalid response header".to_string()));
+        }
+
+        let seq = u64::from_be_bytes([buf[0], buf[1], buf[2], buf[3], buf[4], buf[5], buf[6], buf[7]]);
+        let op = buf[8];
+        let len = u64::from_be_bytes([buf[9], buf[10], buf[11], buf[12], buf[13], buf[14], buf[15], buf[16]]);
+
+        Ok(RespHeader { seq, op, len })
+    }
+}
+
+/// Define the Packet trait, for once send or receive packets
+///
+/// Client will create a packet and serialize it to a byte array,
+/// send it to the server. And create a temp response packet in client.
+///
+/// Server will receive the packet, deserialize it to a packet struct,
+/// and process the request, then create a response packet and serialize.
+///
+/// Client will receive the response packet and deserialize it to a packet struct.
+/// and check the status of the packet and the response.
+pub trait Packet: Sync + Send + Clone + Debug {
+    /// Get the packet seq number
+    fn seq(&self) -> u64;
+    /// Set the packet seq number
+    fn set_seq(&mut self, seq: u64);
+
+    /// Get packet type
+    fn op(&self) -> u8;
+    /// Set packet type
+    fn set_op(&mut self, op: u8);
+
+    /// Serialize request data to bytes
+    fn set_req_data(&mut self, data: &[u8]) -> Result<(), RpcError<String>>;
+
+    /// Get the serialized request data
+    fn get_req_data(&self) -> Result<Vec<u8>, RpcError<String>>;
+
+    /// Serialize response data to bytes
+    fn set_resp_data(&mut self, data: &[u8]) -> Result<(), RpcError<String>>;
+
+    /// Deserialize response data from bytes
+    fn get_resp_data(&self) -> Result<Vec<u8>, RpcError<String>>;
+
+    /// Get the packet status
+    fn status(&self) -> PacketStatus;
+
+    /// Set the packet status
+    fn set_status(&mut self, status: PacketStatus);
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum PacketStatus {
```

Comment:
This can be replace with Option<Result<(), ErrorType>>

Reply:
Ok, I have fixed it.